### ISSUE TITLE
Keycloak gatekeeper 500 Internal Server Error

### ISSUE THREAD

**Issue by 8ctv** [louketo/louketo-proxy#624]
Hi, I'm using keycloak gatekeeper in Openshift to connect to kibana. Users are complaining that sometimes they get error 500 Internal Server Error while performing searches. I was able to reproduce the error, as seen below. 

The requests don't reach kibana from gatekeeper whenever the error occurs. 

At first I thought that it's related to the kc-access token that expires in 300s, but I could get the error in the time range as well. Also, you can see below in the pasted logs that the kc-access token expires in xx seconds, but after it expires, gatekeeper keeps using it ("expires": -157.261452856). This is not neccesarily the root cause of the issue as I could get searches going beyond 600s after expiration (-600s).

Image: docker.io/keycloak/keycloak-gatekeeper:5.0.0

Command:
```
/opt/keycloak-gatekeeper
--upstream-url=http://kibana:5601
--enable-refresh-tokens=true
--enable-login-handler=true
--enable-metrics=true
--enable-logging=true
--resources=uri=/*
--discovery-url=https://keycloak/auth/realms/openshift
--secure-cookie=false
--client-id=openshift
--client-secret=secret
--listen=0.0.0.0:8080
--encryption-key=key
--redirection-url=https://kibana
--skip-openid-provider-tls-verify=true
--upstream-response-header-timeout=300s
--server-read-timeout=300s
--server-write-timeout=300s
--cors-origins="*"
--cors-methods=PUT
--cors-methods=POST
--cors-methods=GET
--max-idle-connections=50
--max-idle-connections-per-host=50
--verbose
```

Keycloak-gatekeeper logs with verbosity enabled:

```
1.5904883049769607e+09  debug   keycloak-gatekeeper/session.go:51       found the user identity {"id": "723rqt8r-5t2r-9097-122q-u918t6vr580q", "name": "demo@example.com", "email": "demo@example.com", "roles": "offline_access,uma_authorization,user,account:manage-account,account:manage-account-links,account:view-profile", "groups": ""}
1.5904883049772062e+09  debug   keycloak-gatekeeper/middleware.go:337   access permitted to resource    {"access": "permitted", "email": "demo@example.com", "expires": 273.022794534, "resource": "/*"}
1.5904883049798055e+09  info    keycloak-gatekeeper/middleware.go:90    client request  {"latency": 0.002943718, "status": 304, "bytes": 0, "client_ip": "10.128.0.1:40676", "method": "GET", "path": "/ui/fonts/roboto_mono/RobotoMono-Regular.ttf"}

1.590488306385567e+09   debug   keycloak-gatekeeper/session.go:51       found the user identity {"id": "723rqt8r-5t2r-9097-122q-u918t6vr580q", "name": "demo@example.com", "email": "demo@example.com", "roles": "offline_access,uma_authorization,user,account:manage-account,account:manage-account-links,account:view-profile", "groups": ""}
1.5904883063858542e+09  debug   keycloak-gatekeeper/middleware.go:337   access permitted to resource    {"access": "permitted", "email": "demo@example.com", "expires": 271.614146805, "resource": "/*"}

1.5904883063863587e+09  debug   keycloak-gatekeeper/session.go:51       found the user identity {"id": "723rqt8r-5t2r-9097-122q-u918t6vr580q", "name": "demo@example.com", "email": "demo@example.com", "roles": "offline_access,uma_authorization,user,account:manage-account,account:manage-account-links,account:view-profile", "groups": ""}
1.590488306386638e+09   debug   keycloak-gatekeeper/middleware.go:337   access permitted to resource    {"access": "permitted", "email": "demo@example.com", "expires": 271.613362844, "resource": "/*"}
1.5904883063898976e+09  info    keycloak-gatekeeper/middleware.go:90    client request  {"latency": 0.004509888, "status": 304, "bytes": 0, "client_ip": "10.128.0.1:40676", "method": "GET", "path": "/built_assets/dlls/icon.arrow_left-js.bundle.dll.js"}
1.5904883063998766e+09  info    keycloak-gatekeeper/middleware.go:90    client request  {"latency": 0.013591594, "status": 304, "bytes": 0, "client_ip": "10.128.0.1:40670", "method": "GET", "path": "/built_assets/dlls/icon.arrow_right-js.bundle.dll.js"}

1.5904883064022453e+09  debug   keycloak-gatekeeper/session.go:51       found the user identity {"id": "723rqt8r-5t2r-9097-122q-u918t6vr580q", "name": "demo@example.com", "email": "demo@example.com", "roles": "offline_access,uma_authorization,user,account:manage-account,account:manage-account-links,account:view-profile", "groups": ""}
1.590488306402439e+09   debug   keycloak-gatekeeper/middleware.go:337   access permitted to resource    {"access": "permitted", "email": "demo@example.com", "expires": 271.597561446, "resource": "/*"}
1.590488306404104e+09   info    keycloak-gatekeeper/middleware.go:90    client request  {"latency": 0.002042336, "status": 304, "bytes": 0, "client_ip": "10.128.0.1:40782", "method": "GET", "path": "/built_assets/dlls/icon.play-js.bundle.dll.js"}

1.590488308136064e+09   debug   keycloak-gatekeeper/session.go:51       found the user identity {"id": "723rqt8r-5t2r-9097-122q-u918t6vr580q", "name": "demo@example.com", "email": "demo@example.com", "roles": "offline_access,uma_authorization,user,account:manage-account,account:manage-account-links,account:view-profile", "groups": ""}
1.59048830813631e+09    debug   keycloak-gatekeeper/middleware.go:337   access permitted to resource    {"access": "permitted", "email": "demo@example.com", "expires": 269.86369069, "resource": "/*"}
1.5904883081423037e+09  info    keycloak-gatekeeper/middleware.go:90    client request  {"latency": 0.006402345, "status": 304, "bytes": 0, "client_ip": "10.128.0.1:40782", "method": "GET", "path": "/built_assets/dlls/icon.kql_function-js.bundle.dll.js"}

1.5904883082380908e+09  debug   keycloak-gatekeeper/session.go:51       found the user identity {"id": "723rqt8r-5t2r-9097-122q-u918t6vr580q", "name": "demo@example.com", "email": "demo@example.com", "roles": "offline_access,uma_authorization,user,account:manage-account,account:manage-account-links,account:view-profile", "groups": ""}
1.5904883082383094e+09  debug   keycloak-gatekeeper/middleware.go:337   access permitted to resource    {"access": "permitted", "email": "demo@example.com", "expires": 269.76169142, "resource": "/*"}
1.5904883115628932e+09  info    keycloak-gatekeeper/middleware.go:90    client request  {"latency": 3.324970429, "status": 200, "bytes": 0, "client_ip": "10.128.0.1:40782", "method": "POST", "path": "/elasticsearch/_msearch"}

1.590488317384172e+09   debug   keycloak-gatekeeper/session.go:51       found the user identity {"id": "723rqt8r-5t2r-9097-122q-u918t6vr580q", "name": "demo@example.com", "email": "demo@example.com", "roles": "offline_access,uma_authorization,user,account:manage-account,account:manage-account-links,account:view-profile", "groups": ""}
1.5904883173844943e+09  debug   keycloak-gatekeeper/middleware.go:337   access permitted to resource    {"access": "permitted", "email": "demo@example.com", "expires": 260.61550668, "resource": "/*"}

1.5904884014486518e+09  debug   keycloak-gatekeeper/session.go:51       found the user identity {"id": "723rqt8r-5t2r-9097-122q-u918t6vr580q", "name": "demo@example.com", "email": "demo@example.com", "roles": "offline_access,uma_authorization,user,account:manage-account,account:manage-account-links,account:view-profile", "groups": ""}
1.5904884014495242e+09  debug   keycloak-gatekeeper/middleware.go:337   access permitted to resource    {"access": "permitted", "email": "demo@example.com", "expires": 176.550477196, "resource": "/*"}
1.5904884014544284e+09  info    keycloak-gatekeeper/middleware.go:90    client request  {"latency": 0.006040168, "status": 500, "bytes": 4, "client_ip": "10.128.0.1:42988", "method": "POST", "path": "/elasticsearch/_msearch"}

1.5904884015075245e+09  debug   keycloak-gatekeeper/session.go:51       found the user identity {"id": "723rqt8r-5t2r-9097-122q-u918t6vr580q", "name": "demo@example.com", "email": "demo@example.com", "roles": "offline_access,uma_authorization,user,account:manage-account,account:manage-account-links,account:view-profile", "groups": ""}
1.5904884015078192e+09  debug   keycloak-gatekeeper/middleware.go:337   access permitted to resource    {"access": "permitted", "email": "demo@example.com", "expires": 176.492181818, "resource": "/*"}
1.5904884015348818e+09  info    keycloak-gatekeeper/middleware.go:90    client request  {"latency": 0.027521626, "status": 304, "bytes": 0, "client_ip": "10.128.0.1:42988", "method": "GET", "path": "/built_assets/dlls/icon.alert-js.bundle.dll.js"}

1.5904884015610573e+09  debug   keycloak-gatekeeper/session.go:51       found the user identity {"id": "723rqt8r-5t2r-9097-122q-u918t6vr580q", "name": "demo@example.com", "email": "demo@example.com", "roles": "offline_access,uma_authorization,user,account:manage-account,account:manage-account-links,account:view-profile", "groups": ""}
1.5904884015613143e+09  debug   keycloak-gatekeeper/middleware.go:337   access permitted to resource    {"access": "permitted", "email": "demo@example.com", "expires": 176.438686679, "resource": "/*"}
1.5904884015648546e+09  info    keycloak-gatekeeper/middleware.go:90    client request  {"latency": 0.004019088, "status": 304, "bytes": 0, "client_ip": "10.128.0.1:42996", "method": "GET", "path": "/built_assets/dlls/icon.cross-js.bundle.dll.js"}

1.5904884160201492e+09  debug   keycloak-gatekeeper/session.go:51       found the user identity {"id": "723rqt8r-5t2r-9097-122q-u918t6vr580q", "name": "demo@example.com", "email": "demo@example.com", "roles": "offline_access,uma_authorization,user,account:manage-account,account:manage-account-links,account:view-profile", "groups": ""}
1.5904884160204484e+09  debug   keycloak-gatekeeper/middleware.go:337   access permitted to resource    {"access": "permitted", "email": "demo@example.com", "expires": 161.979552587, "resource": "/*"}
1.5904885570995383e+09  info    keycloak-gatekeeper/middleware.go:90    client request  {"latency": 239.715538453, "status": 200, "bytes": 0, "client_ip": "10.128.0.1:40782", "method": "POST", "path": "/elasticsearch/_msearch"}
1.5904885612524924e+09  info    keycloak-gatekeeper/middleware.go:90    client request  {"latency": 145.232518125, "status": 200, "bytes": 0, "client_ip": "10.128.0.1:42996", "method": "POST", "path": "/elasticsearch/_msearch"}

1.5904887350906627e+09  debug   keycloak-gatekeeper/session.go:51       found the user identity {"id": "723rqt8r-5t2r-9097-122q-u918t6vr580q", "name": "demo@example.com", "email": "demo@example.com", "roles": "offline_access,uma_authorization,user,account:manage-account,account:manage-account-links,account:view-profile", "groups": ""}
1.5904887350909033e+09  info    keycloak-gatekeeper/middleware.go:154   accces token for user has expired, attemping to refresh the token       {"client_ip": "10.128.0.1:51526", "email": "demo@example.com"}
1.590488735261247e+09   info    keycloak-gatekeeper/middleware.go:190   injecting the refreshed access token cookie     {"client_ip": "10.128.0.1:51526", "cookie_name": "kc-access", "email": "demo@example.com", "expires_in": 299.738760859}
1.590488735261454e+09   debug   keycloak-gatekeeper/middleware.go:337   access permitted to resource    {"access": "permitted", "email": "demo@example.com", "expires": -157.261452856, "resource": "/*"}
1.5904887352634256e+09  info    keycloak-gatekeeper/middleware.go:90    client request  {"latency": 0.173288616, "status": 500, "bytes": 4, "client_ip": "10.128.0.1:51526", "method": "POST", "path": "/elasticsearch/_msearch"}

1.5904888926389794e+09  debug   keycloak-gatekeeper/session.go:51       found the user identity {"id": "723rqt8r-5t2r-9097-122q-u918t6vr580q", "name": "demo@example.com", "email": "demo@example.com", "roles": "offline_access,uma_authorization,user,account:manage-account,account:manage-account-links,account:view-profile", "groups": ""}
1.59048889263938e+09    debug   keycloak-gatekeeper/middleware.go:337   access permitted to resource    {"access": "permitted", "email": "demo@example.com", "expires": 142.360620948, "resource": "/*"}
1.590488909058796e+09   info    keycloak-gatekeeper/middleware.go:90    client request  {"latency": 16.420258418, "status": 200, "bytes": 0, "client_ip": "10.128.0.1:55642", "method": "POST", "path": "/elasticsearch/_msearch"}
```

**Comment by jangaraj** [comment]
That looks like Kibana `/elasticsearch/_msearch` returns 500 and then this error is just returned back to the client. It really doesn't look like gatekeeper generates 500.

You may try to use newer image, e.g. 
```
quay.io/keycloak/keycloak-gatekeeper:10.0.0
```

**Comment by 8ctv** [comment]
I could not get the error while bypassing keycloak gatekeeper using port forward directly on kibana. 
Also, whenever the error occurs I could not see any logs coming from the kibana pod which is why I was quite sure that the issue lied in the gatekeeper.

Thank you @jangaraj for suggesting the image from quay.io. I'm trying for 30 minutes now to get the error, with no luck :)

I remember I used it before but got an empty page with a Temporary Redirect link on the Kibana site and I didn't bother with it, assuming it was broken, when all I had to do was delete the related session info from my browser.
